src/miniscript/evaluate.rs

```diff
@@ -0,0 +1,561 @@
+fn evaluate_time(n: u32, age: u32, witness: &mut Vec<Vec<u8>>) -> Result<Primitives, Error>
+{
+    let top = witness.pop().expect("Witness Stack must contain one element");
+    if top == vec![]{
+        witness.push(vec![]);
+        Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![]})
+    }
+    else if top == vec![1]{
+        let res = evaluate_csv(n, age)?;
+        witness.push(vec![1]);
+        Ok(res)
+    }
+    else {
+        Err(Error::CouldnotEvaluate)
+    }
+}
+
+/// helper function for calculating equivalent of HASH256 hash OP_EQUAL
+/// len(preimage) = 32 and SHA256(preimage) = hash
+fn evaluate_hash(hash: sha256::Hash, preimage: &Vec<u8>, witness: &mut Vec<Vec<u8>>)
+                   -> Result<Primitives, Error >
+{
+    if preimage.len() != 32{
+        return Err(Error::VerifyHashFail(hash, preimage.clone()))
+    }
+    if sha256::Hash::hash(preimage) == hash{
+        witness.push(vec![1]);
+        Ok(Primitives{pks: vec![], hashlocks: vec![preimage.clone()], timelocks: vec![]})
+    }
+    else{
+        Err(Error::VerifyHashFail(hash, preimage.clone()))
+    }
+}
+
+fn evaluate_hash_t(hash: sha256::Hash, witness:&mut Vec<Vec<u8>>)
+    -> Result<Primitives, Error >
+{
+    let preimage = witness.pop().expect("Preimage missing");
+    evaluate_hash(hash, &preimage, witness)
+}
+
+fn evaluate_hash_w(hash: sha256::Hash, witness: &mut Vec<Vec<u8>>)
+    -> Result<Primitives, Error >
+{
+    let witnesslen = witness.len();
+    witness.swap(witnesslen-1, witnesslen -2);//swap the last elements
+    let preimage = witness.pop().expect("Preimage missing");
+    if preimage == vec![]{
+        witness.push(vec![]);
+        return Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![]})
+    }
+    evaluate_hash(hash, &preimage, witness)
+}
+
+/// A and B. Evaluates both A and B and then computes BOOLAND
+fn interpret_and_bool<P>(
+    left: &AstElem<P>,
+    right: &AstElem<P>,
+    witness: &mut Vec<Vec<u8>>,
+    age: u32,
+    msg: &secp256k1::Message,
+    secp: &secp256k1::Secp256k1<All>)
+    -> Result<Primitives, Error >
+    where P: ToPublicKey + Clone,
+{
+    let mut res = left.interpret(witness, age, msg, secp)?;
+    res.join(&right.interpret(witness, age, msg, secp)?);
+
+    let leftres = witness.pop().expect("Witness stack has at least one element");
+    let rightres = witness.pop().expect("Witness stack has at least one element");
+    let andres = leftres == vec![1] && rightres == vec![1];
+    if andres{
+        witness.push(vec![1]);
+    }
+    else{
+        //Even if the script fragment evaluates to false, this method will return the pks, hash
+        // and timelocks used
+        witness.push(vec![]);
+    }
+    Ok(res)
+}
+
+/// If !A then 0 else B. Evaluates B only if A is true
+fn interpret_and_casc<P>(
+    left: &AstElem<P>,
+    right: &AstElem<P>,
+    witness: &mut Vec<Vec<u8>>,
+    age: u32,
+    msg: &secp256k1::Message,
+    secp: &secp256k1::Secp256k1<All>)
```

Comment:
Nit: I'd prefer `secp` be the first argument of every function it appears in, rather than the last. You should also use a `Secp256k1<C>` with `C: secp256k1::Verification` rather than `Secp256k1<All>`.

Reply:
In general I think the order of arguments should be
1. `secp`
2. `stack` (renamed from `witness`)
3. `sighash` (renamed from `msg`)
4. `age`
5. everything else (in this case `left` and `right`)

Reply:
Changed